### PR TITLE
fix(passkeys): serialize WebAuthn credentials without native toJSON()

### PR DESCRIPTION
**Pull request by vpomerleau**
## Because

- On iOS/WebKit < 26.5.1, creating a passkey in Mozilla Accounts settings enters an infinite "setup" loop: the WebContent renderer crashes (an uncatchable process trap) inside the native `PublicKeyCredential.toJSON()` extension serializer whenever a PRF extension output is present, so the page reloads and re-attempts endlessly.
- The serializer that triggers the crash isn't needed — the long-standing WebAuthn Level 2 getters produce the same JSON without the broken code path.

## This pull request

- Replaces native `PublicKeyCredential.toJSON()` with a hand-rolled serializer (`toCredentialJSON`/`serializeResponse` in `packages/fxa-settings/src/lib/passkeys/webauthn.ts`) that reads the L2 getters (`rawId`, `response.*`, `getClientExtensionResults()`) on both registration and authentication.
- Replaces the Level 3 parse-helper feature gate with a Level 2 baseline check (`isWebAuthnSupported`) and decodes the backend's request options to native `ArrayBuffer`s by hand (`toNativeCreationOptions`/`toNativeRequestOptions`), broadening passkey support to all WebAuthn-L2 browsers (iOS 16+).
- Extracts shared base64url helpers to `packages/fxa-settings/src/lib/base64url.ts`, used by the serializer and the WebAuthn helper storybook (fixing a latent non-chunked encoder bug there).
- Syncs the functional-test WebAuthn polyfill (`packages/functional-tests/lib/passkeyPolyfill.ts`) to the L2-getter/serialize contract and hardens it to reject malformed native options; adds a Chromium CDP virtual-authenticator smoke test (`packages/functional-tests/tests/passkeyAuth/passkey-cdp.spec.ts`) exercising the real WebAuthn stack end-to-end.

## Issue that this pull request solves

Issue: FXA-13991 (iOS half; the Windows desktop fix is a separate branch, PR #20786 — this does not close the shared ticket)

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).
- [x] I have manually reviewed all AI generated code.

## How to review (Optional)

- Key files/areas to focus on: `webauthn.ts` (manual serializer + manual request decoders + `isWebAuthnSupported`), `base64url.ts`, `passkeyPolyfill.ts` (test-harness sync + hardening), `passkey-cdp.spec.ts`.
- Suggested review order: `base64url.ts` → `webauthn.ts` (serializer → decoders → gate) → `passkeyPolyfill.ts` → `passkey-cdp.spec.ts`.
- Risky or complex parts: the serializer must never call native `toJSON()` (the crash); the decoders must decode exactly the base64url fields (challenge, user.id, credential ids, PRF salts). The server cryptographically re-verifies challenge/origin/signature, so a serialize/decode bug fails the ceremony rather than silently corrupting data.

## Screenshots (Optional)

Please attach the screenshots of the changes made in case of change in user interface.

## Other information (Optional)

- **iOS half of FXA-13991.** The Windows desktop fix (static PRF salt + silent no-PRF retry) is a separate branch — PR #20786. Footer is `Issue:` (not `Closes:`) so this doesn't auto-resolve the shared ticket. `webauthn.ts` will conflict-merge with that branch (complementary changes).
- **Broader than the crash fix.** Dropping the L3 parse-helper gate for a Level 2 baseline broadens passkey support to all WebAuthn-L2 browsers (iOS 16+, older Chrome/Firefox); the previous L3 JSON-helper gate excluded everything before Safari/iOS 18.4, Chrome 129, Firefox 119.
- **Manual iOS QA still required.** The functional polyfill and the Chromium CDP virtual-authenticator test validate protocol/crypto + API shape, but neither reproduces WebKit-specific behaviour (the iOS `toJSON()` crash itself) — real Safari/iOS device QA remains the backstop. The CDP smoke test is tagged `@chromium` and runs only under the local Playwright `-chromium` projects — it is not run in CI, so treat it as a local/manual QA aid.
- **Functional test timeout fix (separate commit).** Trimmed the passwordless passkey AAL2 divert test (`passkey-signin.spec.ts`) to stop at the `/inline_totp_setup` divert; it was timing out at 60s+ by re-running the full inline-TOTP enrolment that the cheaper passkey-only divert test already covers. Now passes locally in ~10s.

### CODE REVIEW

File: packages/functional-tests/lib/cdpVirtualAuthenticator.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      // Best-effort teardown; the context may already be closing.
+      try {
+        await client.send('WebAuthn.removeVirtualAuthenticator', {
+          authenticatorId,
+        });
+        await client.detach();
+      } catch {
+        // ignore
+      }
+    },
```
[the remark this answers is not in the document]
Comment (vpomerleau) [reply packages/functional-tests/lib/cdpVirtualAuthenticator.ts:61]: Fixed — `remove()` now runs `removeVirtualAuthenticator` and `detach()` as independent best-effort steps, so the CDP session is detached even if removal throws.

File: packages/fxa-settings/src/lib/passkeys/webauthn.ts [hunk 2]
```diff
@@ -206,13 +296,24 @@ function toCredentialJSON(
...
+    response: serializeResponse(credential.response, operation),
+    // Passed through as-is. Binary extension outputs (e.g. PRF results) stay
+    // ArrayBuffers and are dropped when the credential is JSON-stringified for
+    // the server — intentional: the server reads only prf.enabled, and the PRF
+    // output must remain client-side (kB stays server-blind).
+    clientExtensionResults:
+      typeof credential.getClientExtensionResults === 'function'
+        ? (credential.getClientExtensionResults() as Record<string, unknown>)
+        : {},
+  };
```
[the remark this answers is not in the document]
Comment (vpomerleau) [reply packages/fxa-settings/src/lib/passkeys/webauthn.ts:292]: Fixed — `toCredentialJSON` now validates the credential shape (`type === 'public-key'`, `rawId` present, `response` a non-null object) and throws `DOMException('UnknownError')` for non-WebAuthn / malformed objects, instead of a bare TypeError from the field reads. (Deliberately no `instanceof ArrayBuffer` on rawId — unreliable across realms; the encoder handles the buffer.)

File: packages/functional-tests/README.md [hunk 3]
```diff
@@ -14,12 +14,14 @@ The environments that this suite may run against are:
...
 Each has its own named script in [package.json](./package.json) or you can use `--project` when running `playwright test` manually. They are implemented in `lib/targets`.
 
-There is also a very small subset of tests that can _only_ run against Chromium browsers. These have slightly different project names, listed below, and should only be used when running the Passkey tests.
+There is also a very small subset of tests that can _only_ run against Chromium browsers (e.g. passkey specs that drive Chrome's real WebAuthn stack via a CDP virtual authenticator). These run under the Chromium-only project variants below:
 
 - local-chromium
 - stage-chromium
 - production-chromium
 
+These specs are tagged `@chromium`, which is the separation mechanism: the Firefox projects set `grepInvert: /@chromium/` (so they skip them) and the `-chromium` projects set `grep: /@chromium/` (so they run _only_ them). Tag any new Chromium-only test's `describe`/`test` title with `@chromium`. **These projects are local/manual only — CI does not run them**, so treat `@chromium` specs as a manual QA aid rather than a CI gate.
```
Comment (nshirley) [inline packages/functional-tests/README.md]: I like the pattern here to tag the tests! I'd suggest having the format match `#<filter>` similar to others though, just for consistency. super nit, though

File: packages/fxa-settings/src/lib/base64url.ts
Comment (nshirley) [inline packages/fxa-settings/src/lib/base64url.ts:1]: It looks like `packages/fxa-auth-client/lib/utils.ts` has similar functionality, and I wonder if that can be used instead of making a new helper here?

File: packages/functional-tests/tests/passkeyAuth/passkey-cdp.spec.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+/*
+ * Chromium-only passkey smoke test: drives Chrome's REAL WebAuthn stack via a CDP
+ * virtual authenticator (see cdpVirtualAuthenticator.ts), not the polyfill, to
+ * exercise the production webauthn.ts path end-to-end.
+ *
+ * A CDP authenticator is still Chromium's WebAuthn, so it can't reproduce the
+ * iOS toJSON() renderer crash — that remains manual cross-device QA.
+ *
+ * Tagged @chromium so it runs only under the -chromium Playwright projects.
```
Comment (nshirley) [inline packages/functional-tests/tests/passkeyAuth/passkey-cdp.spec.ts]: Probably don't need the comment stating it's tagged with this, a bit redundant 😄

File: packages/functional-tests/playwright.config.ts [hunk 5]
```diff
@@ -85,6 +88,9 @@ export default defineConfig<PlaywrightTestConfig<TestOptions, WorkerOptions>>({
     ),
     ...(TargetNames.map((name) => ({
       name: `${name}-chromium`,
+      // Run ONLY @chromium-tagged specs (passkey CDP virtual-authenticator
```
Comment (nshirley) [inline packages/functional-tests/playwright.config.ts]: imo, the comments feel unnecessary, it's pretty clear what's happening with `grep` and `grepInvert`

Comment (nshirley) [reply packages/functional-tests/playwright.config.ts]: I see it's already noted/covered in the README too, so definitely redundant

File: packages/functional-tests/README.md (hunk 3 above)
Comment (nshirley) [reply packages/functional-tests/README.md]: To piggy back a bit
> **These projects are local/manual only — CI does not run them**
Why is that? I'm pretty sure I probably made that decision when adding the first attempt at these tests... so maybe I'm just forgetting why 😅

Comment (vpomerleau) [reply packages/functional-tests/README.md]: IIRC, running on CI was disabled before you joined - and mainly because we were running the entire suite on tests on chromium and that was heavy. We could revisit running a few target tests like these ones, perhaps?

👍 about the tag update, I'll take care of that

File: packages/fxa-settings/src/lib/base64url.ts
Comment (vpomerleau) [reply packages/fxa-settings/src/lib/base64url.ts:1]: They are similar but not identical. The webauthn specific helpers differ in two main ways:
- chunked encode (spread overflows on a multi-KB `attestationObject`)
- re-padding on decode (server sends unpadded `base64url`)

I'd rather limit the scope to the webauthn helper rather than modifying the original helpers in a way that isn't needed there. WDYT?

File: packages/functional-tests/README.md (hunk 3 above)
Comment (nshirley) [reply packages/functional-tests/README.md]: Cool, that makes sense, thanks for the info!

File: packages/fxa-settings/src/lib/base64url.ts
Comment (nshirley) [reply packages/fxa-settings/src/lib/base64url.ts:1]: That's fair, and I suppose that existing helper is part of the auth-client so it's probably scoped to that. Probably wouldn't make a lot of sense to pull that in.

File: packages/functional-tests/tests/passkeyAuth/passkey-cdp.spec.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+      testAccountTracker,
+    }) => {
+      const authenticator = await addVirtualAuthenticator(page);
+      try {
+        // Sign up + password sign-in to reach Settings.
+        const credentials = await testAccountTracker.signUp();
+        await page.goto(target.contentServerUrl);
+        await signin.fillOutEmailFirstForm(credentials.email);
+        await signin.fillOutPasswordForm(credentials.password);
+        await page.waitForURL(/settings/);
```
Comment (nshirley) [inline packages/functional-tests/tests/passkeyAuth/passkey-cdp.spec.ts:45]: This is probably a "me" issue, so I'm looking further, but I just wanted to see if I could run this test against local and it fails at this step. The page is waiting for an emailed 2FA code, but the test is expecting we're already on settings. I'm assuming you don't get the same thing?

Comment (nshirley) [reply packages/functional-tests/tests/passkeyAuth/passkey-cdp.spec.ts:45]: nvm, it was a "me" issue 🤦 I had `SIGNIN_CONFIRMATION_SKIP_FOR_NEW_ACCOUNTS` set to false in my .zshrc

**Review by nshirley — APPROVED** [review]
lgtm! I ran through testing locally and ran both the polyfill tests and real tests to make sure those all worked - my comment on the test was just a bad setup on my env
